### ISSUE TITLE
Rely on pytest-asyncio capabilities

### ISSUE THREAD

**Issue by argaen** [aio-libs/aiohttp#2496]
I think it would be healthy to rely some of the asynchronous testing capabilities to https://github.com/pytest-dev/pytest-asyncio/.

For example it provides support for asynchronous fixtures, provides an `event_loop` fixture, etc.

IMO `pytest-aiohttp` should dedicate to improve/provide fixtures related to aiohttp itself like `test_client` and the likes and let the other things more asyncio related to `pytest-asyncio`. This way we would also avoid problems (https://github.com/pytest-dev/pytest-asyncio/issues/52) where users use `pytest-asyncio` and `pytest-aiohttp` for testing (myself included)

If you agree I volunteer to start a MR to see how it would look like.

**Comment by asvetlov** [comment]
What is MR?

**Comment by argaen** [comment]
Pull Request/Merge Request. Github/Gitlab terminology :)

**Comment by asvetlov** [comment]
You could try but I expect pitfalls on this way.

**Comment by argaen** [comment]
I'll have a look at it and see how it goes

**Comment by hynek** [comment]
As far as I can tell, currently the testing facilities of aiohttp (like aiohttp_client) are incompatible with pytest-asyncio. I get errors about missing event_loops which suggests that they screw up each others setups:

```
========================================================================= ERRORS =========================================================================
_____________________________________________ ERROR at setup of TestPool.test_remove_from_pool_simple[None] ______________________________________________

reader = <StreamReader>, writer = FakeWriter(_buf=b'', _closed=False), event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
timeouts = BackendConfig.Registry.Timeouts(connect=10, request=100, keep_alive=10)

    @pytest.fixture
    def client(reader, writer, event_loop, timeouts):
        """
        A future of a functional raw client.
        """
>       connector = asyncio.Future()

conftest.py:120:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.pyenv/versions/3.6.4/lib/python3.6/asyncio/events.py:694: in get_event_loop
    return get_event_loop_policy().get_event_loop()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <asyncio.unix_events._UnixDefaultEventLoopPolicy object at 0x10f5cb240>

    def get_event_loop(self):
        """Get the event loop.

            This may be None or an instance of EventLoop.
            """
        if (self._local._loop is None and
            not self._local._set_called and
            isinstance(threading.current_thread(), threading._MainThread)):
            self.set_event_loop(self.new_event_loop())
        if self._local._loop is None:
            raise RuntimeError('There is no current event loop in thread %r.'
>                              % threading.current_thread().name)
E           RuntimeError: There is no current event loop in thread 'MainThread'.

../../.pyenv/versions/3.6.4/lib/python3.6/asyncio/events.py:602: RuntimeError
```

As you can see this happens even though I’m not using the fixtures at once.

This is a super bummer because one of the reasons I love to using async frameworks is the option of easily mixing a web framework with general purpose network code. While I’m sure interop can be created in other ways, it would be really nice if we had one really good goto solution that everyone relied on. :)

**Comment by d21d3q** [comment]
According to #1097, well written app allows to passing loop explicitly on init.
I wrote my doubts in #1069 but since that issue is closed, I'll repeat them here.
So why loop argument in Application class is deprecated and `_set_loop` method is prefixed with `_` (which suggests that it is meant for internal use)?
(Edit: Is it only for deprecation period, and in future it will become loop setter?)

I've managed to get rid of `'There is no current event loop in thread %r.'` error, by having `app.py` like this:
```python
import asyncio
from aiohttp.web import Application, run_app

class Foo:
    def __init__(self, app):
        self.queue = asyncio.Queue(loop=app.loop)  # this will crush with pytest without current loop

def make_app(loop=None):
    app = Application()
    if loop is not None:
        app._set_loop(loop)

    app.foo = Foo(app)  # this one will get loop from app 

if __name__ == '__main__':
    run_app(make_app())
```
and in `conftest.py`:
```python
import pytest
from myapp import make_app

@pytest.fixture
def cli(loop, aiohttp_client):
    return loop.run_until_complete(aiohttp_client(make_app(loop)))
```

**Comment by asvetlov** [comment]
What pytest plugin do you use for running async tests? `pytest-asyncio` or `pytest-aiohttp`?

**Comment by d21d3q** [comment]
I have both installed. Not all tests are related to `aiohttp`.

**Comment by hynek** [comment]
Yeah I think this is a key point: if I write a “normal” web app, I just go for Falcon or Pyramid. I specifically use aiohttp only if the application has a good reason to be async. That usually means that it communicates  with the world in other ways than just HTTP.

That’s why aiohttp's testing facilities remain useless to me as long as they stop me from testing the usually more important parts of my applications.

**Comment by asvetlov** [comment]
Well, simple question: What pytest-asyncio feature is missing in pytest-aiohttp?

**Comment by hynek** [comment]
Excellent question! I kind of assumed it’s lacking something but it doesn’t seem to. It even saves me the `@pytest.mark.asnycio` marker! I just converted my test suite and everything seems to be working fine, so I think I’m good!

**Comment by asvetlov** [comment]
Nice to her it.
`pytest-asyncio` is a pretty good name, `pytest-aiohttp` should work with it eventually.
But it is not the first priority because `pytest-aiohttp` just works.
Say again `pytest-asyncio` and `pytest-aiohttp` used very different approaches for the loop life-cicle and other low details but now the difference is not such bug (but still exists).

I'll manage a smooth transition someday but keeping backward compatibility is a pain as usual.

**Comment by ghuname** [comment]
On https://us-pycon-2019-tutorial.readthedocs.io/aiohttp_tests.html is written:
Note: pytest-aiohttp is not compatible with another pupular plugin pytest-asyncio.
Please can you confirm that this is not the case now?

**Comment by asvetlov** [comment]
IMHO it is, still.

**Comment by Dreamsorcerer** [comment]
First time looking at this thread, but reading through I notice there is also just some bad asyncio code being attributed to a problem with the pytest plugins...

For example:

> ```python
> class Foo:
>     def __init__(self, app):
>         self.queue = asyncio.Queue(loop=app.loop)  # this will crush with pytest without current loop
> 
> def make_app(loop=None):
>     app = Application()
>     if loop is not None:
>         app._set_loop(loop)
> 
>     app.foo = Foo(app)  # this one will get loop from app 
> 
> if __name__ == '__main__':
>     run_app(make_app())
> ```

`asynico.Queue()` should (and in 3.10 probably needs) to be called in a running loop. But, the code tries to instantiate it in a normal function before starting the loop (which happens in `run_app`). The correct approach would be to instantiate this on startup, something like:

> ```python
> class Foo:
>     def __init__(self):
>         self.queue = asyncio.Queue()
>
> async def startup(app):
>     app.foo = Foo()
> 
> def make_app():
>     app = Application()
>     app.on_startup.append(startup)
> 
> if __name__ == '__main__':
>     run_app(make_app())
> ```

**Comment by ghuname** [comment]
> IMHO it is, still.

Is there a simple example how to mock an aiohttp get request?
For example, how to write pytest for RestClient.get_employees()? What if endpoints are class's or method's parameters? I would like to mock session.get('https://dummy.restapiexample.com/public/api/v1/employees') in order to produce fixed json as an output.

```
class Employees(BaseModel):
    all_employees: Dict
    single_employee: Dict


class RestClient:

    async def get_employees(self, employee_id: int) -> Employees:
        async with aiohttp.ClientSession() as session:
            t1 = asyncio.create_task(session.get('https://dummy.restapiexample.com/public/api/v1/employees'))
            t2 = asyncio.create_task(
                session.get(f'https://dummy.restapiexample.com/public/api/v1/employee/{employee_id}'))

            # all employees json: {"status":"success","data":[{"id":1,"employee_name":"Tiger Nixon","employee_salary":320800,"employee_age":61,"profile_image":""},{"id":2,"employee_name":"Garrett Winters","employee_salary":170750,"employee_age":63,"profile_image":""},{"id":3,"employee_name":"Ashton Cox","employee_salary":86000,"employee_age":66,"profile_image":""},{"id":4,"employee_name":"Cedric Kelly","employee_salary":433060,"employee_age":22,"profile_image":""},{"id":5,"employee_name":"Airi Satou","employee_salary":162700,"employee_age":33,"profile_image":""},{"id":6,"employee_name":"Brielle Williamson","employee_salary":372000,"employee_age":61,"profile_image":""},{"id":7,"employee_name":"Herrod Chandler","employee_salary":137500,"employee_age":59,"profile_image":""},{"id":8,"employee_name":"Rhona Davidson","employee_salary":327900,"employee_age":55,"profile_image":""},{"id":9,"employee_name":"Colleen Hurst","employee_salary":205500,"employee_age":39,"profile_image":""},{"id":10,"employee_name":"Sonya Frost","employee_salary":103600,"employee_age":23,"profile_image":""},{"id":11,"employee_name":"Jena Gaines","employee_salary":90560,"employee_age":30,"profile_image":""},{"id":12,"employee_name":"Quinn Flynn","employee_salary":342000,"employee_age":22,"profile_image":""},{"id":13,"employee_name":"Charde Marshall","employee_salary":470600,"employee_age":36,"profile_image":""},{"id":14,"employee_name":"Haley Kennedy","employee_salary":313500,"employee_age":43,"profile_image":""},{"id":15,"employee_name":"Tatyana Fitzpatrick","employee_salary":385750,"employee_age":19,"profile_image":""},{"id":16,"employee_name":"Michael Silva","employee_salary":198500,"employee_age":66,"profile_image":""},{"id":17,"employee_name":"Paul Byrd","employee_salary":725000,"employee_age":64,"profile_image":""},{"id":18,"employee_name":"Gloria Little","employee_salary":237500,"employee_age":59,"profile_image":""},{"id":19,"employee_name":"Bradley Greer","employee_salary":132000,"employee_age":41,"profile_image":""},{"id":20,"employee_name":"Dai Rios","employee_salary":217500,"employee_age":35,"profile_image":""},{"id":21,"employee_name":"Jenette Caldwell","employee_salary":345000,"employee_age":30,"profile_image":""},{"id":22,"employee_name":"Yuri Berry","employee_salary":675000,"employee_age":40,"profile_image":""},{"id":23,"employee_name":"Caesar Vance","employee_salary":106450,"employee_age":21,"profile_image":""},{"id":24,"employee_name":"Doris Wilder","employee_salary":85600,"employee_age":23,"profile_image":""}],"message":"Successfully! All records has been fetched."}
            # single employee json {"status":"success","data":{"id":1,"employee_name":"Tiger Nixon","employee_salary":320800,"employee_age":61,"profile_image":""},"message":"Successfully! Record has been fetched."}

            t1_resp, t2_resp = await asyncio.gather(t1, t2)
            t1_dict, t2_dict = await asyncio.gather(t1_resp.json(), t2_resp.json())

        return Employees(
            all_employees=t1_dict,
            single_employee=t2_dict
        )


async def main():
    client = RestClient()
    return await client.get_employees(employee_id=1)


if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    print(loop.run_until_complete(main()))

```

Output:

`all_employees={'status': 'success', 'data': [{'id': 1, 'employee_name': 'Tiger Nixon', 'employee_salary': 320800, 'employee_age': 61, 'profile_image': ''}, {'id': 2, 'employee_name': 'Garrett Winters', 'employee_salary': 170750, 'employee_age': 63, 'profile_image': ''}, {'id': 3, 'employee_name': 'Ashton Cox', 'employee_salary': 86000, 'employee_age': 66, 'profile_image': ''}, {'id': 4, 'employee_name': 'Cedric Kelly', 'employee_salary': 433060, 'employee_age': 22, 'profile_image': ''}, {'id': 5, 'employee_name': 'Airi Satou', 'employee_salary': 162700, 'employee_age': 33, 'profile_image': ''}, {'id': 6, 'employee_name': 'Brielle Williamson', 'employee_salary': 372000, 'employee_age': 61, 'profile_image': ''}, {'id': 7, 'employee_name': 'Herrod Chandler', 'employee_salary': 137500, 'employee_age': 59, 'profile_image': ''}, {'id': 8, 'employee_name': 'Rhona Davidson', 'employee_salary': 327900, 'employee_age': 55, 'profile_image': ''}, {'id': 9, 'employee_name': 'Colleen Hurst', 'employee_salary': 205500, 'employee_age': 39, 'profile_image': ''}, {'id': 10, 'employee_name': 'Sonya Frost', 'employee_salary': 103600, 'employee_age': 23, 'profile_image': ''}, {'id': 11, 'employee_name': 'Jena Gaines', 'employee_salary': 90560, 'employee_age': 30, 'profile_image': ''}, {'id': 12, 'employee_name': 'Quinn Flynn', 'employee_salary': 342000, 'employee_age': 22, 'profile_image': ''}, {'id': 13, 'employee_name': 'Charde Marshall', 'employee_salary': 470600, 'employee_age': 36, 'profile_image': ''}, {'id': 14, 'employee_name': 'Haley Kennedy', 'employee_salary': 313500, 'employee_age': 43, 'profile_image': ''}, {'id': 15, 'employee_name': 'Tatyana Fitzpatrick', 'employee_salary': 385750, 'employee_age': 19, 'profile_image': ''}, {'id': 16, 'employee_name': 'Michael Silva', 'employee_salary': 198500, 'employee_age': 66, 'profile_image': ''}, {'id': 17, 'employee_name': 'Paul Byrd', 'employee_salary': 725000, 'employee_age': 64, 'profile_image': ''}, {'id': 18, 'employee_name': 'Gloria Little', 'employee_salary': 237500, 'employee_age': 59, 'profile_image': ''}, {'id': 19, 'employee_name': 'Bradley Greer', 'employee_salary': 132000, 'employee_age': 41, 'profile_image': ''}, {'id': 20, 'employee_name': 'Dai Rios', 'employee_salary': 217500, 'employee_age': 35, 'profile_image': ''}, {'id': 21, 'employee_name': 'Jenette Caldwell', 'employee_salary': 345000, 'employee_age': 30, 'profile_image': ''}, {'id': 22, 'employee_name': 'Yuri Berry', 'employee_salary': 675000, 'employee_age': 40, 'profile_image': ''}, {'id': 23, 'employee_name': 'Caesar Vance', 'employee_salary': 106450, 'employee_age': 21, 'profile_image': ''}, {'id': 24, 'employee_name': 'Doris Wilder', 'employee_salary': 85600, 'employee_age': 23, 'profile_image': ''}], 'message': 'Successfully! All records has been fetched.'} single_employee={'status': 'success', 'data': {'id': 1, 'employee_name': 'Tiger Nixon', 'employee_salary': 320800, 'employee_age': 61, 'profile_image': ''}, 'message': 'Successfully! Record has been fetched.'}`

**Comment by Dreamsorcerer** [comment]
> Is there a simple example how to mock an aiohttp get request?

Not related to this report, so please open a new discussion next time to avoid spamming others.

asvetlov usually seems to recommend creating a test server, which avoids mocking entirely, and allows to fully test the connection. I often prefer mocking though, in which case you can take a look at the aioresponses library.

**Comment by ghuname** [comment]
Sorry for not related post. I tried to open a new issue, but this picture intimidated me:

![image](https://user-images.githubusercontent.com/15207713/147972494-82ea681e-7fc1-4956-922f-a2afb9fc0f72.png)

I don't know where to put it.
Anyway, I can delete the post.

**Comment by merwok** [comment]
You have a usage question, so you can ask on StackOverflow or Discourse.
If you need something that is not provided by the module: Feature request
If it’s there but not clearly documented: Bug report

**Comment by ghuname** [comment]
I tried SO but with no success. 
Anyway I am deleting my post.

**Comment by merwok** [comment]
Sorry to hear that!  Note that Dreamsorcerer gave you an answer: mocking is one possible technique (see `aioresponses` on PyPI), running a local HTTP server to have real requests is another

**Comment by webknjaz** [comment]
@merwok

> You have a usage question, so you can ask on StackOverflow or Discourse.

FWIW we should probably replace the Discourse link with https://github.com/aio-libs/aiohttp/discussions (the org-global link is already pointing at https://github.com/aio-libs/.github/discussions).

@ghuname sounds like you may want to start a conversation at https://github.com/aio-libs/aiohttp/discussions/categories/q-a.

**Comment by ghuname** [comment]
Thanks, I have started discussion there on https://github.com/aio-libs/aiohttp/discussions/6497

**Comment by asvetlov** [comment]
Fixed by pytest-asyncio 0.17 and pytest-aiohttp 1.0
